library/core/src/primitive_docs.rs

```diff
@@ -1317,6 +1317,45 @@ mod prim_f16 {}
 /// | `wasm32`, `wasm64` | If all input NaNs are quiet with all-zero payload: None.<br> Otherwise: all possible payloads. |
 ///
 /// For targets not in this table, all payloads are possible.
+///
+/// # Algebraic operators
+///
+/// Algebraic operators of the form `a.algebraic_*(b)` allow the compiler to optimize
+/// floating point operations using all the usual algebraic properties of real numbers --
+/// despite the fact that those properties do *not* hold on floating point numbers.
+/// This can give a great performance boost since it may unlock vectorization.
+///
+/// The exact set of optimizations is unspecified but typically allows combining operations,
+/// rearranging series of operations based on mathematical properties, converting between division and reciprocal multiplication, and
+/// disregarding the sign of zero. This means that the results of elementary operations may have
+/// undefined precision, but will not invoke input-dependent undefined behavior.
+///
+/// Because of the unpredictable nature of compiler optimizations, the same inputs may produce
+/// different results even within a single program run. **Unsafe code must not rely on any property
+/// of the return value for soundness.** However, implementations will generally do their best to
+/// pick a reasonable tradeoff between performance and accuracy of the result.
+///
+/// For example:
+///
+/// ```
+/// # #![feature(float_algebraic)]
+/// # #![allow(unused_assignments)]
+/// # let mut x: f32 = 12.0;
+/// # let a: f32 = 34.0;
+/// # let b: f32 = 56.0;
+/// x = x.algebraic_add(a.algebraic_mul(b));
+/// ```
+///
+/// May be rewritten as either:
+///
+/// ```
+/// # #![allow(unused_assignments)]
+/// # let mut x: f32 = 12.0;
+/// # let a: f32 = 34.0;
+/// # let b: f32 = 56.0;
+/// x = x + (a * b); // As written
+/// x = (a * b) + x; // Reordered to allow using a single `fma`
```

Comment:
I think this example is misleading since fp addition (and multiplication) are commutative (the rewriting is not significant) and it's the `AllowContract` flag that is necessary to allow use of fmuladd (when the target feature is available).

An alternate example would be `a + b + c + d` being rewritten as `(a + c) + (b + d)`, which has a shorter critical path and allows packed vector instruction to do `[a, b] + [c, d]`. If you prefer to illustrate with `AllowContract`, one could use `a * d - b * c` to be rewritten as `fmuladd(a, d, - b*c)`, which is only two instructions (but will produce nonzero if a=b, c=d).

Reply:
Thanks! Updated, how's this look?